text/0016-generalized-entity-store.md

```diff
@@ -0,0 +1,113 @@
+# Generalized Entity Store
+
+## Related issues and PRs
+
+- Implementation PR(s): https://github.com/cedar-policy/cedar/pull/171, which is a pull request from https://github.com/prakol16/cedar
+
+## Timeline
+
+- Start Date: 2023-07-11
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Entity storage ideally should not be restricted to JSON files. We should either allow users to add their own custom storage methods or add support for some non-JSON common storage methods (e.g. databases) ourselves.
+
+## Basic example
+
+
+```rust
+struct MyOwnDatabase {
+    // ...
+}
+
+impl EntityDatabase for MyOwnDatabase {
+    fn get_entity_of_uid(&self, uid: &EntityUID) -> Option<Entity> {
+        // custom implementation which could e.g. make a request to a database
+    }
+}
+```
+
+This could be code written by users, or we could hide `EntityDatabase` and instead, there could be several storage solutions (other than the default JSON one) that we write ourselves.
+
+## Motivation
+
+  - Currently, every request reads the entire JSON file, reparsing all of the attributes, even though the overwhelming majority of requests will need to read only a few entries of the entity store. In general, with many entities, it may not even be feasible to load all of the entities into memory at once.
+
+  - In the future, we can turn partial expression evaluations into queries so that we can efficiently find e.g. all of the photos that a user has access to. Ideally, we would leverage existing query languages such as SQL and existing storage solutions which are already optimized to respond to these kinds of queries efficiently, rather than try to rebuild this on our own.
+
+  - Some users may already have infrastructure that uses databases or other storage methods for their applications. This makes integrating Cedar into existing software easier, because users are not boxed in to one particular storage format.
+
+
+## Detailed design
+
+
+We would generalize the authorizer to accept any `EntityDatabase`, not just `Entities`:
+
+```rust
+impl Authorizer {
+    // old:
+    pub fn is_authorized_old(&self, q: &Request, pset: &PolicySet, entities: &Entities) -> Response
+
+    // proposed change:
+    pub fn is_authorized<T : EntityDatabase>(&self, q: &Request, pset: &PolicySet, entities: &T) -> Response
+}
+```
+
+Here is the current interface for `EntityDatabase` (see [the source](https://github.com/prakol16/cedar/blob/e5df52e118e378c2142aaa6d9a4781f0f67b8064/cedar-policy-core/src/entities.rs#L265) for more comments and the implementation of the trait for `Entities`):
+
+```rust
+pub trait EntityDatabase {
+    /// Given a uid, get the corresponding entity if it exists
+    /// The entity should already have an `ancestors` attribute which represents
+    /// the transitive closure ancestors, not the direct parents
```

Comment:
Yes, basically it's transferring work from Cedar to the storage engine. The idea is that this computation could be expensive so should be computed once, and updated whenever the underlying entity store is modified. This will result in performance improvements when there are not many updates to the entity store, but the performance will probably be worse if there are many writes to the entity store.

We could compute the TC ourselves; we will have to do a DFS for each entity we retrieve. Each DFS might be very fast, or might result in retrieving all O(n^2) edges of the graph (and note that n, the number of entities, is implicit, so if the user has written a function that implicitly has infinitely many entities, this could potentially not halt), so this makes the timings of the query more unpredictable.

An improvement to consider would be to have the `EntityDatabase` tell us whether we should expect the transitive closure to already be computed or not. We can reuse the `TCComputation` enum to tell us this information. This allows the user to control whether they want the storage to do the tc computation and cache it or they want us to do it. Should I add this to the RFC?

Reply:
> An improvement to consider would be to have the EntityDatabase tell us whether we should expect the transitive closure to already be computed or not. We can reuse the TCComputation enum to tell us this information. This allows the user to control whether they want the storage to do the tc computation and cache it or they want us to do it. Should I add this to the RFC?

I think having the option to have the storage say "yeah nah I'm not going to do that" would be helpful, but I'm not 100% clear on why the engine needs the transitive closure of all ancestors; this feels like a "sometimes" thing to me as a newcomer to Cedar rather than an "always" thing, because I don't necessarily see how the ancestry information is needed in a number of the types of policies I've been thinking about (some of them, absolutely. All of them? Not so much.) Is there something I could read to help me understand why the transitive closure of ancestors is needed?

Reply:
The transitive closure behavior is already in Cedar, and is not new in this PR. See https://docs.cedarpolicy.com/syntax-operators.html#hierarchy-and-set-membership-operators-and-functions

Reply:
The transitive closure is needed to allow efficient handling of the `in` operator. Cedar doesn't force you to include all of the transitive ancestors, for example in (https://github.com/cedar-policy/cedar-examples/tree/main/cedar-example-use-cases/github_example) you can perform entity slicing by passing in the User, the Issue / Repo and any *relevant* ancestors. Suppose `User Bob` can read `Repo 1` because Bob is a member of `Team A` which can read `Repo 1`. Then you'll need to pass in `User Bob`, `Repo 1` and `Team A`. You don't need to include other members of `Team A`, other Teams Bob is a member of or other Repos `Team A` can read. That's for the `github_example`. Doing this requires thinking carefully about what constitutes correct entity slicing for your application. Notably there are no forbid policies in `github_examples`, so you don't have to worry about e.g., omitting Bob's membership of some Team that is forbidden to read `Repo 1`.

Reply:
>a function in EntityDatabase that answers the question "is entity E in entity F` ?

I like this idea. See my comment here: https://github.com/cedar-policy/rfcs/pull/16#discussion_r1275390544 for more.

Reply:
The rewrite in #28 addresses this -- both in its main proposal, and potentially even further in its Alternative J.  Resolving this comment, but please feel free to continue this conversation on #28.